### PR TITLE
field does not exist error: note fields if Levenshtein suggestion fails

### LINKED ISSUE #42599: List the available field names of a struct instead of reporting "field not found"
Rust errors have improved a lot in this direction over the past year, but this would make it even easier to code without having to look for struct field names in the documentation:

```
#[derive(Default)]
struct A { a: (), b: (), c:(), d: (), e: () }

fn main() {
   let a = A::default();
   let b = a.z;
}
```

At the moment, rustc reports "did you mean `a`?". It would be more convenient to report "no such field. Available fields are: a, b, c, d, e".

In a codebase in a "prototyping stage", it can be the case that several types have similar field names. Listing the fields would even make it easier to refactor, by highlighting similarities.

### PR DESCRIPTION
**Pull request by zackmdavis**
When trying to access or initialize a nonexistent field, if we can't infer what
field was meant (by virtue of the purported field in the source being a small
Levenshtein distance away from an actual field, suggestive of a typo), issue a
note listing all the available fields. To reduce terminal clutter, we don't
issue the note when we have a `find_best_match_for_name` Levenshtein
suggestion: the suggestion is probably right.

The third argument of the call to `find_best_match_for_name` is changed to
`None`, accepting the default maximum Levenshtein distance of one-third of the
identifier supplied for correction. The previous value of `Some(name.len())`
was overzealous, inappropriately very Levenshtein-distant suggestions when the
attempted field access could not plausibly be a mere typo. For example, if a
struct has fields `mule` and `phone`, but I type `.donkey`, I'd rather the
error have a note listing that the available fields are, in fact, `mule` and
`phone` (which is the behavior induced by this patch) rather than the error
asking "did you mean `phone`?" (which is the behavior on master). The "only
find fits with at least one matching letter" comment was accurate when it was
first introduced in 09d992471 (January 2015), but is a vicious lie in its
present context before a call to `find_best_match_for_name` and must be
destroyed (replacing every letter is within a Levenshtein distance of name.len()).

The present author claims that this suffices to resolve #42599.

### CODE REVIEW

**Comment by eddyb** [comment]
Is there any way to get @rust-highfive to try randomly assigning again?

r? @nikomatsakis

File: src/librustc_typeck/check/mod.rs [hunk 1]
```diff
@@ -3107,14 +3122,22 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
...
+                    if adt.is_enum() {
+                        err.span_label(field.name.span,
+                                       format!("`{}::{}` does not have this field",
+                                               ty, variant.name));
+                    } else {
+                        err.span_label(field.name.span,
+                                       format!("`{}` does not have this field", ty));
+                    }
+                    let available_field_names = self.available_field_names(variant);
+                    err.note(&format!("available fields are: {}",
```
Comment (nikomatsakis) [inline src/librustc_typeck/check/mod.rs]: same here, of course

File: src/librustc_typeck/check/mod.rs [hunk 2]
```diff
@@ -2956,6 +2956,11 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                                                format!("did you mean `{}`?", suggested_field_name));
                             } else {
                                 err.span_label(field.span, "unknown field");
+                                let struct_variant_def = def.struct_variant();
+                                let available_field_names = self.available_field_names(
+                                    struct_variant_def);
+                                err.note(&format!("available fields are: {}",
+                                                  available_field_names.join(", ")));
```
Comment (nikomatsakis) [inline src/librustc_typeck/check/mod.rs]: hmm; ideally we would put backticks around these field names. But, more importantly, we generally have a rule that we only list out things exhaustively up to a certain limit (typically 5), and then after that we will say "...and others" or something. See e.g. this code [in method suggestions](https://github.com/rust-lang/rust/blob/master/src/librustc_typeck/check/method/suggest.rs#L88).

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
This looks great! I hate silly suggestions. But there are some stylistic nits.

**Comment by zackmdavis** [comment]
2dbfa39 adds the backticks, "... and _n_ others" truncation, and declines to issue the note if we can't actually see any fields

**Review by nikomatsakis — APPROVED** [review]
Looks great, thanks!

**Pull request merged**